### PR TITLE
Add warnings for cgroup related options (`--cgroup*`) when `--pod` or `--pod-id-file` is set

### PR DESCRIPTION
**Pull request by ruihe774**
#### Does this PR introduce a user-facing change?

```release-note
Added warnings for cgroup related options (`--cgroup*`) when `--pod` or `--pod-id-file` is set.
```

### CODE REVIEW

**Comment by ruihe774** [comment]
cc @rhatdan

File: cmd/podman/containers/create.go [hunk 1]
```diff
@@ -299,12 +299,31 @@ func CreateInit(c *cobra.Command, vals entities.ContainerCreateOptions, isInfra
...
-		if c.Flag("cgroups").Changed && vals.CgroupsMode == "split" && registry.IsRemote() {
+
+		cgroupsChanged := c.Flag("cgroups").Changed
+		if cgroupsChanged && vals.CgroupsMode == "split" && registry.IsRemote() {
 			return vals, fmt.Errorf("the option --cgroups=%q is not supported in remote mode", vals.CgroupsMode)
 		}
 
-		if c.Flag("pod").Changed && !strings.HasPrefix(c.Flag("pod").Value.String(), "new:") && c.Flag("userns").Changed {
-			return vals, errors.New("--userns and --pod cannot be set together")
+		if c.Flag("userns").Changed {
```
Comment (mheon) [inline cmd/podman/containers/create.go:308]: I wonder if we should add this to Libpod's `container_validate.go` to make sure that, even if we add additional flags, setting a user namespace while joining a pod is not allowed

Comment (ruihe774) [reply cmd/podman/containers/create.go:308]: I have no idea how to distinguish normal containers and infra containers at libpod level. The latter can have userns set.

Comment (mheon) [reply cmd/podman/containers/create.go:308]: `ctr.config.IsInfra` bool should do it

**Comment by ruihe774** [comment]
Giving setting `--userns` and `--pod` together does not cause serious problem, I have changed my mind to make it a warning instead of error.

File: cmd/podman/containers/create.go (hunk 1 above)
Comment (ruihe774) [reply cmd/podman/containers/create.go]: I have just changed it to a warning in CLI.

**Comment by mheon** [comment]
I thought there was a reason the two were incompatible... @giuseppe @Luap99 Do you remember what it was?

**Comment by giuseppe** [comment]
we must use the same user namespace for all the containers in a pod, otherwise we can't share namespaces.

**Comment by ruihe774** [comment]
> we must use the same user namespace for all the containers in a pod, otherwise we can't share namespaces.

There are already many ways to break this assumption. For example, set `--cgroups` when creating containers.

**Comment by ruihe774** [comment]
> we must use the same user namespace for all the containers in a pod, otherwise we can't share namespaces.

I cannot find this guarantee in doc. With different user namespace the pod do not crash and runs well. I think a warning is enough. Also, it's difficult to ensure this guarantee.

**Comment by giuseppe** [comment]
> > we must use the same user namespace for all the containers in a pod, otherwise we can't share namespaces.
> 
> I cannot find this guarantee in doc. With different user namespaces the pod do not crash and runs well. I think a warning is enough. Also, it's difficult to ensure this guarantee.

if you don't share any namespace, why would you need to use a pod?

**Comment by rhatdan** [comment]
Using a Pod allows you to treat the service as a single application.  I don't see a requirement to share namespaces.

**Comment by giuseppe** [comment]
> Using a Pod allows you to treat the service as a single application. I don't see a requirement to share namespaces.

in that case we need to enchance the check and verify that the list of shared namespaces is empty.  By default pods share the ipc, net, and uts namespaces, so something like: `podman pod create foo && podman run --pod foo --userns ...` won't work

**Comment by rhatdan** [comment]
Yes I think you would need to change the namespaces shared on the created pod to "".

**Comment by ruihe774** [comment]
> > Using a Pod allows you to treat the service as a single application. I don't see a requirement to share namespaces.
> 
> in that case we need to enchance the check and verify that the list of shared namespaces is empty. By default pods share the ipc, net, and uts namespaces, so something like: `podman pod create foo && podman run --pod foo --userns ...` won't work

I think it is too complicated to implement such checks. And IMO sharing these namespaces or not is  orthogonal to each other.

**Comment by Luap99** [comment]
Doing these checks in the front end is fundamentally flawed. We simply do not have enough knowledge there to know what namespaces are used together and what conflicts,etc... These checks really must be moved into specgen instead, because cli level checks do not cover the API.

And overall I agree that we should not throw warnings if the namespace is not shared in the first place, this would be easy to know in specgen.

> I cannot find this guarantee in doc. With different user namespaces the pod do not crash and runs well

Well if you do not care about the "owning" user namespaces sure. User namespaces are special in terms of how they work with capabilities and permissions in general, i.e. https://blog.podman.io/2023/12/interaction-between-user-namespaces-and-capabilities/
So yes it can work thus I agree the hard error always seemed a bit harsh but on the other side for most users they will not understand these details which will lead to confusing issues.

**Comment by ruihe774** [comment]
> Doing these checks in the front end is fundamentally flawed. We simply do not have enough knowledge there to know what namespaces are used together and what conflicts,etc... These checks really must be moved into specgen instead, because cli level checks do not cover the API.
>
> And overall I agree that we should not throw warnings if the namespace is not shared in the first place, this would be easy to know in specgen.

I'm not familiar with the codebase. Making such change is too difficult for me. Maybe I shall drop the userns commit out of this PR and only keep the cgroups one. Hope that someone can work on that.

**Review by Luap99 — COMMENTED** [review]
IMO this is still the wrong place to throw such a warning even for cgroups.

If the pod was created without any cgroup limits then it now always throws warnings to users that didn't need it to begin with. I think structurally it is important to only throw warnings when they are actual warnings, i.e. unexpected things.
Just consider than any quadlet unit using pods will not throw these warnings by default even when no limit was set on the pod which just doesn't make sense for users.

Overall we can only check these things in specgen. The current solution will not work for any API user, it will not work when defaults are set via config file as you only check the flags.

**Comment by ruihe774** [comment]
> Overall we can only check these things in specgen. The current solution will not work for any API user, it will not work when defaults are set via config file as you only check the flags.

The checks were orignally placed here. I just extended them. Despite that, I agree with you that they should be moved into specgen, which is out of my knowledge.

> If the pod was created without any cgroup limits then it now always throws warnings to users that didn't need it to begin with. I think structurally it is important to only throw warnings when they are actual warnings, i.e. unexpected things.

I do not agree. IMO users *expect* cgroups of containers are descendants of the cgroup of the pod.

**Pull request closed without merging**